File: algorithm/JavaScript/SuperUglyNumber.js
/**
 * @param {number} n
 * @param {number[]} primes
 * @return {number}
 */
 var nthSuperUglyNumber = function(n, primes) {
  const dp = new Array(n + 1).fill(0);
  dp[1] = 1;
  const m = primes.length;
  const pointers = new Array(m).fill(1);
  for (let i = 2; i <= n; i++) {
      const nums = new Array(m).fill(m);
      let minNum = Number.MAX_SAFE_INTEGER;
      for (let j = 0; j < m; j++) {
          nums[j] = dp[pointers[j]] * primes[j];
          minNum = Math.min(minNum, nums[j]);
      }
      dp[i] = minNum;
      for (let j = 0; j < m; j++) {
          if (minNum == nums[j]) {
              pointers[j]++;
          }
      }

  }
  return dp[n];
};

File: algorithm/python/SuperUglyNumber.py
class Solution:
    def nthSuperUglyNumber(self, n, primes):
        dp = [0 for _ in range(n+1)]
        dp[1] = 1
        pointers = [1 for _ in range(len(primes))]

        for i in range(2, n+1):
            min_num = min(pointers[j] * primes[j] for j in range(len(primes)))
            dp[i] = min_num
            for j in range(len(primes)):
                if dp[pointers[j]] * primes[j] == min_num:
                    pointers[j] += 1
        
        return dp[n]